### PR TITLE
Add State Generator to DB

### LINKED ISSUE #3063: Eliminate Usage of HistoricalStates - Use Exponential Backoff for State Generation
Hi all,

An important task for becoming production ready is preventing the usage of HistoricalStates throughout our repo. Currently, we store every single state to the DB, which is not sustainable in a real environment. Instead, we should have the ability to regenerate states through only storing very few of them since the last finalized epoch. This is a tracking issue so we do not forget this important piece of functionality.

### PR DESCRIPTION
**Pull request by 0xKiwi**
Resolves #3063 

This PR is for the removal of HistoricalStates to be replaced with a set interval to save states, and a state generator to use the saved blocks to regenerate the state at a requested slot.  

This adds a new constant to the config `SavingInterval` to declare how often we will save states. The default for this is 8 slots but it can be easily changed as we optimize.

Every 8 slots, the state is saved, if there is a skipped slot over the saving interval slot, then we save after the next processed block.
If there is a skip larger than `SavingInterval`, then we recursively check the DB until we find a state that should be saved. This should rarely happen but this is designed with that in mind.

This design is laid out [here](https://docs.google.com/document/d/19WvJuVc3fAPdO4R79pjo3FkjZ5-FkBSQmQPAIuON9Fc/edit?usp=sharing).

This is a benchmark for processing 8 slots to regenerate a requested state.
```
BenchmarkGenerateStateAtSlot_WorstCase-4   	      10	 273289656 ns/op	220067684 B/op	 1485374 allocs/op
```

### CODE REVIEW

File: validator/client/validator_propose.go [hunk 1]
```diff
@@ -55,31 +55,31 @@ func (v *validator) ProposeBlock(ctx context.Context, slot uint64, pk string) {
...
 	}
-	root, err := ssz.SigningRoot(b)
+	root, err := ssz.SigningRoot(block)
 	if err != nil {
 		log.WithError(err).Error("Failed to sign block")
 		return
 	}
 	signature := v.keys[pk].SecretKey.Sign(root[:], domain.SignatureDomain)
-	b.Signature = signature.Marshal()
+	block.Signature = signature.Marshal()
```
Comment (terencechain) [inline validator/client/validator_propose.go]: Why change it? 'b' was fine there? didn't see any warning on that

Comment (0xKiwi) [reply validator/client/validator_propose.go]: `b` is a vague name, figured it would be a good idea to change it.

File: shared/testutil/block.go [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/state"
+	pb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+	ethpb "github.com/prysmaticlabs/prysm/proto/eth/v1alpha1"
+	"github.com/prysmaticlabs/prysm/shared/bls"
+	"github.com/prysmaticlabs/prysm/shared/params"
+	"github.com/prysmaticlabs/prysm/shared/trieutil"
+)
+
+// GenerateFullBlock generates a fully valid block with the requested parameters.
+func GenerateFullBlock(
```
Comment (0xKiwi) [inline shared/testutil/block.go]: This is a util for generating fully valid blocks, if you would like me to move this into a different PR (with it's own tests as well), that would be fine.

File: validator/client/validator_propose.go (hunk 1 above)
Comment (terencechain) [reply validator/client/validator_propose.go]: b is good for local variable, it's close to what go standard recommended. This is a great reference: https://talks.golang.org/2014/names.slide#6

e.g. r for reader, w for writer, etc

Comment (terencechain) [reply validator/client/validator_propose.go]: Most time we use block instead of b is because b is already used as one of the package name

Comment (0xKiwi) [reply validator/client/validator_propose.go]: Ahhh I see, I'll info it then

**Comment by nisdas** [comment]
what is the status on this ? @0xKiwi

**Comment by 0xKiwi** [comment]
> what is the status on this ? @0xKiwi

Apologies for letting it go stale, still need to test it in runtime. Unsure on how to test it properly since one main case to test for is reorgs.

**Comment by 0xKiwi** [comment]
Closing since I have higher priority tasks to handle and this isn't ready yet.
